README
NAME
    Getopt::Attribute - Attribute wrapper for Getopt::Long

VERSION
    version 2.101410

SYNOPSIS
      use Getopt::Attribute;

      our $verbose : Getopt(verbose!);
      our $all     : Getopt(all);
      our $size    : Getopt(size=s);
      our $more    : Getopt(more+);
      our @library : Getopt(library=s);
      our %defines : Getopt(define=s);
      sub quiet : Getopt(quiet) { our $quiet_msg = 'seen quiet' }
      usage() if our $man : Getopt(man);

      # Meanwhile, on some command line:
      #
      # mypgm.pl --noverbose --all --size=23 --more --more --more --quiet
      #          --library lib/stdlib --library lib/extlib
      #          --define os=linux --define vendor=redhat --man -- foo

DESCRIPTION
    Note: This version of the module works works with perl 5.8.0. If you
    need it to work with perl 5.6.x, please use an earlier version from
    CPAN.

    This module provides an attribute wrapper around "Getopt::Long". Instead
    of declaring the options in a hash with references to the variables and
    subroutines affected by the options, you can use the "Getopt" attribute
    on the variables and subroutines directly.

    As you can see from the Synopsis, the attribute takes an argument of the
    same format as you would give as the hash key for "Getopt::Long". See
    the "Getopt::Long" manpage for details.

    Note that since attributes are processed during CHECK, but assignments
    on newly declared variables are processed during run-time, you can't set
    defaults on those variables beforehand, like this:

            our $verbose : Getopt(verbose!) = 1;  # DOES NOT WORK

    Instead, you have to establish defaults afterwards, like so:

            our $verbose : Getopt(verbose!);
            $verbose ||= 1;

    Alternatively, you can specify a default value within the "Getopt"
    attribute:

        our $def2 : Getopt(def2=i 42);

    To check whether there was an error during "getopt" processing you can
    use the "error()" function:

        pod2usage(-verbose => 2, -exitval => 0) if Getopt::Attribute->error;

METHODS
  error
    FIXME

  options
    FIXME

INSTALLATION
    See perlmodinstall for information and options on installing Perl
    modules.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=Getopt-Attribute>.

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see
    <http://search.cpan.org/dist/Getopt-Attribute/>.

    The development version lives at
    <http://github.com/hanekomu/Getopt-Attribute/>. Instead of sending
    patches, please fork this project using the standard git and github
    infrastructure.

AUTHOR
      Marcel Gruenauer <marcel@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2001 by Marcel Gruenauer.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Getopt-Attribute

2.101410  2010-05-21 12:38:39 Europe/Vienna
    - set major version to 2 so version continuity on CPAN is preserved.

1.100860  2010-03-27 11:23:21 Europe/Vienna
    - converted the distribution to Dist::Zilla-style

1.46  Mon 2009.12.07 00:53:55 CET (Marcel Gruenauer <marcel@cpan.org>)
    - Changed dist style and Changes back to standard. People didn't like it -
      the nail that sticks out gets hammered down.
    - Added standard test files; this will also help with CPANTS.
    - reduced dependencies using author tests
    - better dist engineering (INSTALL, README, README.mkdn etc.)

1.45  Thu, 29 May 2008 12:05:12 -0000 (Marcel Gruenauer <marcel@cpan.org>)
    - Converted Changes file to YAML style
    - .shipit: added Twitter step
    - Makefile.PL: added auto_install() and process_templates()
    - lib/*: converted to template
    - updated MANIFEST
    - added t/perlcriticrc
    - updated MANIFEST
    - .shipit: fixed svk.tagpattern
    - tags: NEWFEATURE

1.44  Thu, 08 Nov 2007 12:09:16 +0100 (Marcel Gruenauer <marcel@cpan.org>)
    - added error()

1.43  Thu, 18 Oct 2007 10:09:39 +0200 (Marcel Gruenauer <marcel@cpan.org>)
    - fixed version requirement of Test::More

1.42  Sun, 09 Sep 2007 23:11:11 +0200 (Marcel Gruenauer <marcel@cpan.org>)
    - require perl 5.8, as I got some test failures on 5.6.x

1.41  Sat, 08 Sep 2007 11:10:40 +0200 (Marcel Gruenauer <marcel@cpan.org>)
    - converted to new-style distribution (ShipIt, Module::Install)
    - moved files to proper locations (lib/, t/)
    - fixed documentation

1.32  Sat, 31 May 2003 18:44:12 +0100 (Marcel Gruenauer <marcel@cpan.org>)
    - added tests

1.30  Thu, 18 Jul 2002 12:37:51 +0100 (Marcel Gruenauer <marcel@cpan.org>)
    - first release version

0.01  Tue, 08 May 2001 04:05:59 +0100 (Marcel Gruenauer <marcel@cpan.org>)
    - original version


Makefile.PL

use strict;
use warnings;

 BEGIN { require 5.008; } 

use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
                       'test' => {
                                   'TESTS' => 't/*.t'
                                 },
                       'NAME' => 'Getopt::Attribute',
                       'DISTNAME' => 'Getopt-Attribute',
                       'CONFIGURE_REQUIRES' => {
                                                 'ExtUtils::MakeMaker' => '6.31'
                                               },
                       'AUTHOR' => 'Marcel Gruenauer <marcel@cpan.org>',
                       'BUILD_REQUIRES' => {
                                             'Test::More' => '0.94',
                                             'Scalar::Util' => '0',
                                             'English' => '0',
                                             'File::Find' => '0',
                                             'File::Temp' => '0',
                                             'Carp' => '0'
                                           },
                       'ABSTRACT' => 'Attribute wrapper for Getopt::Long',
                       'EXE_FILES' => [],
                       'VERSION' => '2.101410',
                       'PREREQ_PM' => {
                                        'Getopt::Long' => '0',
                                        'Attribute::Handlers' => '0'
                                      },
                       'LICENSE' => 'perl'
                     );


delete $WriteMakefileArgs{BUILD_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.56) };

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Getopt/Attribute.pm
use 5.008;
use strict;
use warnings;

package Getopt::Attribute;
BEGIN {
  $Getopt::Attribute::VERSION = '2.101410';
}
# ABSTRACT: Attribute wrapper for Getopt::Long
use Getopt::Long;
use Attribute::Handlers;

sub UNIVERSAL::Getopt : ATTR(RAWDATA,BEGIN) {
    my ($ref, $data) = @_[ 2, 4 ];
    our %options;

    # this has to be an array as we're chasing refs later
    if ($data =~ m/^(\S+)\s+(.*)/) {
        $data = $1;
        push our @defaults, [ $ref => $2 ];
    }
    $options{$data} = $ref;
}
INIT {
    our $error = 0;
    GetOptions(our %options) or $error = 1;
    defined ${ $_->[0] } or ${ $_->[0] } = $_->[1] for our @defaults;
}
sub options { our %options }
sub error   { our $error }
1;


__END__
=pod

=head1 NAME

Getopt::Attribute - Attribute wrapper for Getopt::Long

=head1 VERSION

version 2.101410

=head1 SYNOPSIS

  use Getopt::Attribute;

  our $verbose : Getopt(verbose!);
  our $all     : Getopt(all);
  our $size    : Getopt(size=s);
  our $more    : Getopt(more+);
  our @library : Getopt(library=s);
  our %defines : Getopt(define=s);
  sub quiet : Getopt(quiet) { our $quiet_msg = 'seen quiet' }
  usage() if our $man : Getopt(man);

  # Meanwhile, on some command line:
  #
  # mypgm.pl --noverbose --all --size=23 --more --more --more --quiet
  #          --library lib/stdlib --library lib/extlib
  #          --define os=linux --define vendor=redhat --man -- foo

=head1 DESCRIPTION

Note: This version of the module works works with perl 5.8.0. If you
need it to work with perl 5.6.x, please use an earlier version from CPAN.

This module provides an attribute wrapper around C<Getopt::Long>.
Instead of declaring the options in a hash with references to the
variables and subroutines affected by the options, you can use the
C<Getopt> attribute on the variables and subroutines directly.

As you can see from the Synopsis, the attribute takes an argument
of the same format as you would give as the hash key for C<Getopt::Long>.
See the C<Getopt::Long> manpage for details.

Note that since attributes are processed during CHECK, but assignments
on newly declared variables are processed during run-time, you
can't set defaults on those variables beforehand, like this:

	our $verbose : Getopt(verbose!) = 1;  # DOES NOT WORK

Instead, you have to establish defaults afterwards, like so:

	our $verbose : Getopt(verbose!);
	$verbose ||= 1;

Alternatively, you can specify a default value within the C<Getopt>
attribute:

    our $def2 : Getopt(def2=i 42);

To check whether there was an error during C<getopt> processing you can use
the C<error()> function:

    pod2usage(-verbose => 2, -exitval => 0) if Getopt::Attribute->error;

=head1 METHODS

=head2 error

FIXME

=head2 options

FIXME

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Getopt-Attribute>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Getopt-Attribute/>.

The development version lives at
L<http://github.com/hanekomu/Getopt-Attribute/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2001 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

